difforig perl-current/lib/ExtUtils/Command.pm perl-current/lib/ExtUtils/MM_Any.pm perl-current/lib/ExtUtils/MM_Unix.pm perl-current/lib/ExtUtils/MM_VMS.pm perl-current/lib/ExtUtils/MM_Win32.pm

diff -u perl-current/lib/ExtUtils/Command.pm.orig perl-current/lib/ExtUtils/Command.pm
--- perl-current/lib/ExtUtils/Command.pm.orig	2009-03-24 16:46:58.270375000 +0000
+++ perl-current/lib/ExtUtils/Command.pm	2009-03-24 16:50:22.364125000 +0000
@@ -10,9 +10,9 @@
 require Exporter;
 use vars qw(@ISA @EXPORT @EXPORT_OK $VERSION);
 @ISA       = qw(Exporter);
-@EXPORT    = qw(cp rm_f rm_rf mv cat eqtime mkpath touch test_f test_d chmod
+@EXPORT    = qw(cp rm_f rm_rf mv cat eqtime mkpath touch test_f test_d test_s chmod
                 dos2unix);
-$VERSION = '1.15';
+$VERSION = '1.15_01';
 
 my $Is_VMS   = $^O eq 'VMS';
 my $Is_Win32 = $^O eq 'MSWin32';
@@ -345,6 +345,19 @@
  exit(-s $ARGV[0] ? 0 : 1);
 }
 
+=item test_s
+
+    test_s file
+
+Tests if a file exists and is not empty (size > 0).  I<Exits> with 0 if it does, 1 if it does not.
+
+=cut 
+
+sub test_s
+{
+ exit(-s $ARGV[0] ? 0 : 1);
+}
+
 =item test_d
 
     test_d directory
diff -u perl-current/lib/ExtUtils/MM_Any.pm.orig perl-current/lib/ExtUtils/MM_Any.pm
--- perl-current/lib/ExtUtils/MM_Any.pm.orig	2009-01-26 20:22:08.000000000 +0000
+++ perl-current/lib/ExtUtils/MM_Any.pm	2009-03-24 16:51:37.926625000 +0000
@@ -457,10 +457,10 @@
 #     push @dirs, qw($(INST_ARCHLIB) $(INST_LIB)
 #                    $(INST_BIN) $(INST_SCRIPT)
 #                    $(INST_MAN1DIR) $(INST_MAN3DIR)
-#                    $(INST_LIBDIR) $(INST_ARCHLIBDIR) $(INST_AUTODIR) 
-#                    $(INST_STATIC) $(INST_DYNAMIC) $(INST_BOOT)
+#                    $(INST_LIBDIR) $(INST_ARCHLIBDIR) $(INST_AUTODIR)
+#                    $(INST_STATIC) $(INST_DYNAMIC)
 #                 );
-                  
+
 
     if( $attribs{FILES} ) {
         # Use @dirs because we don't know what's in here.
@@ -643,10 +643,11 @@
 
 sub dynamic {
 # --- Dynamic Loading Sections ---
+# Note:  $(INST_BOOT) moved to dynamic_lib to get rid of empty .bs
 
     my($self) = shift;
     '
-dynamic :: $(FIRST_MAKEFILE) $(INST_DYNAMIC) $(INST_BOOT)
+dynamic :: $(FIRST_MAKEFILE) $(BOOTSTRAP) $(INST_DYNAMIC)
 	$(NOECHO) $(NOOP)
 ';
 }
diff -u perl-current/lib/ExtUtils/MM_Unix.pm.orig perl-current/lib/ExtUtils/MM_Unix.pm
--- perl-current/lib/ExtUtils/MM_Unix.pm.orig	2008-12-19 09:13:57.000000000 +0000
+++ perl-current/lib/ExtUtils/MM_Unix.pm	2009-03-24 16:46:57.864125000 +0000
@@ -864,7 +864,7 @@
 
 # As Mkbootstrap might not write a file (if none is required)
 # we use touch to prevent make continually trying to remake it.
-# The DynaLoader only reads a non-empty file.
+# BOOTSTRAP only gets installed if non-empty.
 $(BOOTSTRAP) : $(FIRST_MAKEFILE) $(BOOTDEP) $(INST_ARCHAUTODIR)$(DFSEP).exists
 	$(NOECHO) $(ECHO) "Running Mkbootstrap for $(NAME) ($(BSLOADLIBS))"
 	$(NOECHO) $(PERLRUN) \
@@ -872,11 +872,6 @@
 		-e "Mkbootstrap('$(BASEEXT)','$(BSLOADLIBS)');"
 	$(NOECHO) $(TOUCH) %s
 	$(CHMOD) $(PERM_RW) %s
-
-$(INST_BOOT) : $(BOOTSTRAP) $(INST_ARCHAUTODIR)$(DFSEP).exists
-	$(NOECHO) $(RM_RF) %s
-	- $(CP) $(BOOTSTRAP) %s
-	$(CHMOD) $(PERM_RW) %s
 MAKE_FRAG
 }
 
@@ -908,7 +903,7 @@
 INST_DYNAMIC_DEP = '.$inst_dynamic_dep.'
 INST_DYNAMIC_FIX = '.$ld_fix.'
 
-$(INST_DYNAMIC): $(OBJECT) $(MYEXTLIB) $(BOOTSTRAP) $(INST_ARCHAUTODIR)$(DFSEP).exists $(EXPORT_LIST) $(PERL_ARCHIVE) $(PERL_ARCHIVE_AFTER) $(INST_DYNAMIC_DEP)
+$(INST_DYNAMIC): $(OBJECT) $(MYEXTLIB) $(INST_ARCHAUTODIR)$(DFSEP).exists $(EXPORT_LIST) $(PERL_ARCHIVE) $(PERL_ARCHIVE_AFTER) $(INST_DYNAMIC_DEP)
 ');
     if ($armaybe ne ':'){
 	$ldfrom = 'tmp$(LIB_EXT)';
@@ -954,8 +949,12 @@
 	  $(INST_DYNAMIC_FIX)
 MAKE
 
+    # copy .bs only if non-empty	
     push @m, <<'MAKE';
 	$(CHMOD) $(PERM_RWX) $@
+	$(NOECHO) $(RM_RF) $(BOOTSTRAP)
+	- $(TEST_S) $(BOOTSTRAP) && $(CP) $(BOOTSTRAP) $(INST_BOOT) && \
+	  $(CHMOD) $(PERM_RW) $(INST_BOOT)
 MAKE
 
     return join('',@m);
@@ -1734,7 +1733,7 @@
 
 Initializes EXTRALIBS, BSLOADLIBS, LDLOADLIBS, LIBS, LD_RUN_PATH, LD,
 OBJECT, BOOTDEP, PERLMAINCC, LDFROM, LINKTYPE, SHELL, NOOP,
-FIRST_MAKEFILE, MAKEFILE_OLD, NOECHO, RM_F, RM_RF, TEST_F,
+FIRST_MAKEFILE, MAKEFILE_OLD, NOECHO, RM_F, RM_RF, TEST_F, TEST_S,
 TOUCH, CP, MV, CHMOD, UMASK_NULL, ECHO, ECHO_N
 
 =cut
@@ -1811,6 +1810,7 @@
     $self->{RM_RF}      ||= "rm -rf";
     $self->{TOUCH}      ||= "touch";
     $self->{TEST_F}     ||= "test -f";
+    $self->{TEST_S}     ||= "test -s";
     $self->{CP}         ||= "cp";
     $self->{MV}         ||= "mv";
     $self->{CHMOD}      ||= "chmod";
@@ -3538,7 +3538,7 @@
 
     # We set PM_FILTER as late as possible so it can see all the earlier
     # on macro-order sensitive makes such as nmake.
-    for my $tool (qw{ SHELL CHMOD CP MV NOOP NOECHO RM_F RM_RF TEST_F TOUCH 
+    for my $tool (qw{ SHELL CHMOD CP MV NOOP NOECHO RM_F RM_RF TEST_F TEST_S TOUCH 
                       UMASK_NULL DEV_NULL MKPATH EQUALIZE_TIMESTAMP 
                       ECHO ECHO_N
                       UNINST VERBINST
diff -u perl-current/lib/ExtUtils/MM_VMS.pm.orig perl-current/lib/ExtUtils/MM_VMS.pm
--- perl-current/lib/ExtUtils/MM_VMS.pm.orig	2008-12-19 09:13:57.000000000 +0000
+++ perl-current/lib/ExtUtils/MM_VMS.pm	2009-03-24 16:46:57.864125000 +0000
@@ -476,6 +476,7 @@
     $self->{RM_F}     ||= '$(ABSPERLRUN) "-MExtUtils::Command" -e rm_f';
     $self->{RM_RF}    ||= '$(ABSPERLRUN) "-MExtUtils::Command" -e rm_rf';
     $self->{TEST_F}   ||= '$(ABSPERLRUN) "-MExtUtils::Command" -e test_f';
+    $self->{TEST_S}   ||= '$(ABSPERLRUN) "-MExtUtils::Command" -e test_s';
     $self->{EQUALIZE_TIMESTAMP} ||= '$(ABSPERLRUN) -we "open F,qq{>>$ARGV[1]};close F;utime(0,(stat($ARGV[0]))[9]+1,$ARGV[1])"';
 
     $self->{MOD_INSTALL} ||= 
diff -u perl-current/lib/ExtUtils/MM_Win32.pm.orig perl-current/lib/ExtUtils/MM_Win32.pm
--- perl-current/lib/ExtUtils/MM_Win32.pm.orig	2008-12-19 09:13:57.000000000 +0000
+++ perl-current/lib/ExtUtils/MM_Win32.pm	2009-03-24 16:46:57.879750000 +0000
@@ -161,6 +161,7 @@
     $self->{MV}       ||= '$(ABSPERLRUN) -MExtUtils::Command -e mv';
     $self->{NOOP}     ||= 'rem';
     $self->{TEST_F}   ||= '$(ABSPERLRUN) -MExtUtils::Command -e test_f';
+    $self->{TEST_S}   ||= '$(ABSPERLRUN) -MExtUtils::Command -e test_s';
     $self->{DEV_NULL} ||= '> NUL';
 
     $self->{FIXIN}    ||= $self->{PERL_CORE} ? 
